app/uk/gov/hmrc/plasticpackagingtaxreturns/services/UserAnswersService.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.plasticpackagingtaxreturns.services
+
+import play.api.mvc.Result
+import play.api.mvc.Results.UnprocessableEntity
+import uk.gov.hmrc.plasticpackagingtaxreturns.models.UserAnswers
+import uk.gov.hmrc.plasticpackagingtaxreturns.repositories.SessionRepository
+import uk.gov.hmrc.plasticpackagingtaxreturns.services.UserAnswersService.notFoundMsg
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+
+class UserAnswersService @Inject()(
+  sessionRepository: SessionRepository
+)(implicit ce: ExecutionContext) {
+
+  def get(cacheKey: String): Future[Either[Result, UserAnswers]] = {
+
+    sessionRepository.get(cacheKey).map(userAnswer => {
+      userAnswer match {
+        case Some(answer) => Right(answer)
+        case _ => Left(UnprocessableEntity(notFoundMsg))
+      }
+    })
+  }
+
+  def get(cacheKey: String, fn: UserAnswers => Future[Result]): Future[Result] = {
```

Comment:
though something like this as the pattern matching from the caller is a quite repetitive. Also have a different flavour (see above) which is used in the AvailableCreditDateRangesController, which left it like that for simplicity.

Reply:
Yep i think this works as a stepping stone for an action or something. Maybe Curry the fn param however so its easy to see the split in the code 
```suggestion
  def get(cacheKey: String)(fn: UserAnswers => Future[Result]): Future[Result] = {
```
which would mean you can use like
```
   userAnswersService.get(request.cacheKey){userAnswers => 
    //blah
    Future(Ok("result"))
  }
```

Reply:
yes, this is good, was looking for something like that. Thanks